mm2src/coins/nft.rs

```diff
@@ -28,102 +29,336 @@ const MORALIS_FORMAT_QUERY_VALUE: &str = "decimal";
+            // this will update only txs related to current nfts in wallet.
+            update_meta_in_txs(&storage, chain, nft_list).await?;
+            update_txs_with_empty_meta(&storage, chain, &req.url).await?;
+        } else {
+            let last_scanned_block = storage.get_last_scanned_block(chain).await?;
+            let last_nft_block = NftListStorageOps::get_last_block_number(&storage, chain).await?;
+
+            match (last_scanned_block, last_nft_block) {
+                // if both block numbers exist, last scanned block should be equal
+                // or higher than last block number from NFT LIST table.
+                (Some(scanned_block), Some(nft_block)) => {
+                    if scanned_block >= nft_block {
+                        update_nft_list(ctx.clone(), &storage, chain, scanned_block + 1, &req.url).await?;
+                        update_txs_with_empty_meta(&storage, chain, &req.url).await?;
+                    } else {
+                        return MmError::err(UpdateNftError::InvalidBlockOrder {
+                            last_scanned_block: scanned_block.to_string(),
+                            last_nft_block: nft_block.to_string(),
+                        });
+                    }
+                },
+                // If the last scanned block value is absent, we cannot accurately update the NFT cache.
+                // This is because a situation may occur where the user doesn't transfer all ERC-1155 tokens,
+                // resulting in the block number of NFT remaining unchanged.
+                (None, Some(nft_block)) => {
+                    return MmError::err(UpdateNftError::LastScannedBlockNotFound {
+                        last_nft_block: nft_block.to_string(),
+                    });
+                },
+                // if there are no rows in NFT LIST table or in both tables there are no rows
+                // we can try to get all info from moralis.
+                (Some(_), None) => {
+                    let nfts = cache_nfts_from_moralis(&ctx, &storage, chain, &req.url).await?;
+                    update_meta_in_txs(&storage, chain, nfts).await?;
+                    update_txs_with_empty_meta(&storage, chain, &req.url).await?;
+                },
+                (None, None) => {
+                    let nfts = cache_nfts_from_moralis(&ctx, &storage, chain, &req.url).await?;
+                    update_meta_in_txs(&storage, chain, nfts).await?;
+                    update_txs_with_empty_meta(&storage, chain, &req.url).await?;
+                },
+            }
+        }
+    }
+    Ok(())
+}
+
+pub async fn refresh_nft_metadata(ctx: MmArc, req: RefreshMetadataReq) -> MmResult<(), UpdateNftError> {
+    let moralis_meta = get_moralis_metadata(
+        format!("{:#02x}", req.token_address),
+        req.token_id.clone(),
+        &req.chain,
+        &req.url,
+    )
+    .await?;
+    let storage = NftStorageBuilder::new(&ctx).build()?;
```

Comment:
do this first in the function, so previously made computing will not be wasted when this fails

Reply:
Well, it is more likely that `get_moralis_metadata` will fail, but its better to avoid sending moralis requests if something else can fail after.
Moved `let storage = NftStorageBuilder::new(&ctx).build()?` to the top.